README.md
# Icalwage - ics to csv

The **icalwage** computes `.ics` file and output `.csv` file.

## usage

```txt
usage: icalwage.py [-h] [-o] [--month MONTH] [--year YEAR] [--from FULL DATE]
                   [--to FULL DATE]
                   ICS FILE CSV FILE

Load calender data from an .ics file, and Output .csv file.

positional arguments:
  ICS FILE          The '.ics' file, this file needed to load work
                    informations.
  CSV FILE          A file name contains csv data.

optional arguments:
  -h, --help        show this help message and exit
  -o, --overwrite   If dest csv file is already exist, overwrite that file.
  --month MONTH     Returns the month data only, use this when ics file is
                    heavy. Please set integer not string. ok: '6' no: 'June'
  --year YEAR       Returns the year data only. Please set christian era. ok:
                    '1989' no: '89'
  --from FULL DATE  If you want to limit the range of a date, use this
                    argument with '--to' arg. Please set full date. ok:
                    '2017.9.1' no: '9.1'
  --to FULL DATE    If you set '--from' argument, please set this argument.
```




CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)

# [Unreleased]

## [v2.1.2] - 2017-7-29

### Fixed

* Fixed date range bug [#1]


## [v2.1.1] - 2017-6-20

### Fixed

* Fixed encoding issue on Windows


## [v2.1.0] - 2017-6-19

### Added

* Added **weekday** item to csv data

### Removed

* Removed needless files


## [v2.0.0] - 2017-6-19

### Changed

* Divided script file
* Changed icalwage subject: **to output csv file only**
* Changed script install approach: to use entry-points
* Changed argument parser to argparse

### Added

* Added **range** arguments: `--month`, `--year`, `--from`, `--to`
* Wroute the unittest

### Fixed

* Compliant pep8


[v2.0.0]: https://github.com/alice1017/icalwage/compare/f33bb2b...v2.0.0
[v2.1.0]: https://github.com/alice1017/icalwage/compare/v2.0.0...v2.1.0
[v2.1.1]: https://github.com/alice1017/icalwage/compare/v2.1.0...v2.1.1
[v2.1.2]: https://github.com/alice1017/icalwage/compare/v2.1.1...v2.1.2
[Unreleased]: https://github.com/alice1017/icalwage/compare/v2.1.2...HEAD

[#1]: https://github.com/alice1017/icalwage/issues/1


tests.py
#!/usr/bin/env python
# coding: utf-8

import pytz
import unittest
import datetime

from icalwage.cli import parser
from icalwage.utils import (
    adjust_args,
    check_destfile,
    check_sourcefile,
    _get_datetime
)


class IcalwageTester(unittest.TestCase):

    def setUp(self):

        pass

    def tearDown(self):

        pass

    def test_adjust(self):

        with self.assertRaises(SystemError):
            args = parser.parse_args(
                [
                    "--year", "1989", "--month", "6",
                    "work.ics", "dest.csv"
                ])
            adjust_args(args)

    def test_adjust2(self):

        with self.assertRaises(SystemError):
            args = parser.parse_args(
                ["--from", "2017.1.1", "work.ics", "dest.csv"])
            adjust_args(args)

    def test_adjust3(self):

        with self.assertRaises(TypeError):
            args = parser.parse_args(
                ["--year", "year", "work.ics", "dest.csv"])
            adjust_args(args)

    def test_adjust4(self):

        with self.assertRaises(TypeError):
            args = parser.parse_args(
                ["--month", "June", "work.ics", "dest.csv"])
            adjust_args(args)

    def test_destfile(self):

        with self.assertRaises(IOError):
            check_destfile("work.ics")

    def test_destfile2(self):

        self.assertTrue(check_destfile("work.ics", overwrite=True))

    def test_sourcefile(self):

        with self.assertRaises(IOError):
            check_sourcefile("source")

    def test_makedtobj(self):

        with self.assertRaises(TypeError):
            _get_datetime("2017")

    def test_makedtobj2(self):

        with self.assertRaises(TypeError):
            _get_datetime("query")

    def test_makedtobj3(self):

        self.assertEqual(
            _get_datetime("2017.1.1"),
            datetime.datetime(
                2017, 1, 1, tzinfo=pytz.timezone("Asia/Tokyo"))
        )

if __name__ == '__main__':

    unittest.main(verbosity=2)


icalwage.py
#!/usr/bin/env python
#coding: utf-8

import os
import sys
import miniparser
from miniparser import kill
from icalendar import Calendar
from dateutil.relativedelta import relativedelta
from StringIO import StringIO

parser = miniparser.parser()
date_format = "%Y/%m/%d"
time_format = "%H:%M:%S"

@parser.option("-s", "--show-event",
        description="Show event list and time. Please set .ics file to arg.")
def show_event(ics_file):
    check_exist_file(ics_file)

    ical = Calendar.from_ical(open(ics_file,"r").read())
    for event in get_events(ical):

        # about time
        start = event.get("dtstart").dt
        end = event.get("dtend").dt
        print start.strftime(date_format), ":",
        print start.strftime(time_format), "~", end.strftime(time_format), ":",

        # about time diff
        delta = relativedelta(end, start)
        print "%d時間" % delta.hours,
        print "%d分"   % delta.minutes

@parser.option("-w", "--work-wage",
        description="Calcuation your wage. Please set ics file and hourly wage",
        argument_types={"ics_file":str, "hourly_wage":int})
def calc_wage(ics_file, hourly_wage):
    check_exist_file(ics_file)

    ical = Calendar.from_ical(open(ics_file,"r").read())

    minutely_wage = hourly_wage/60
    total_wage = 0

    for event in get_events(ical):

        # about time
        start = event.get("dtstart").dt
        end = event.get("dtend").dt
        print start.strftime(date_format), ":",
        print start.strftime(time_format), "~", end.strftime(time_format), ":",

        # about time diff
        delta = relativedelta(end, start)
        total_wage += ((hourly_wage*delta.hours)+(minutely_wage*delta.minutes)) 
        print "%d時間" % delta.hours,
        print ("%d分" if len(str(delta.minutes)) == 2 else "0%d分") % delta.minutes,
        print ":", ((hourly_wage*delta.hours)+(minutely_wage*delta.minutes))
    print "-"*50
    print "総額 :", total_wage

@parser.option("-c", "--csv",
        description="Make CSV file. Please set ics file and file name.")
def make_csv(ics_file, filename):
    check_exist_file(ics_file)

    if os.access(filename, os.F_OK):
        print "fatal: '%s' file is already exist." % filename
        kill(1)

    if filename.endswith(".csv") == False:
        filename = filename+".csv"

    ical = Calendar.from_ical(open(ics_file,"r").read())
    fp = open(filename, "w")
    io = StringIO()
    sys.stdout = io

    put = sys.stdout.write

    for event in get_events(ical):

        # about time
        start = event.get("dtstart").dt
        end = event.get("dtend").dt
        put(start.strftime(date_format) + ",")
        put(start.strftime(time_format) + ",~," + end.strftime(time_format) + ",")

        # about time diff
        delta = relativedelta(end, start)
        put(u"%d時間" % delta.hours)
        put( (u"%d分," if len(str(delta.minutes)) == 2 else u"0%d分,") % delta.minutes )

        # time numberling
        float_min = float(delta.minutes) / 60
        time_num = round(float(delta.hours) + float_min,2)
        put("%g," % time_num)

        # minus 0.5
        put("%g\n" % (round(time_num - 0.5,2)))

    sys.stdout = sys.__stdout__
    fp.write(io.getvalue().encode("SHIFT_JIS"))
    fp.close()

def get_events(calendar):
    for component in calendar.walk():
        if component.name == "VEVENT":
            yield component

def check_exist_file(path):
    if os.access(path, os.F_OK):
        return True
    else:
        print "fatal: file not found."
        kill(1)

if __name__ == "__main__":
    parser.parse()


icalwage/core.py
#!/usr/bin/env python
# coding: utf-8

from icalendar import Calendar
from dateutil.relativedelta import relativedelta

DATE_FORMAT = u"%Y/%m/%d"
TIME_FORMAT = u"%H:%M:%S"
WEEKDAYS = [u"月", u"火", u"水", u"木", u"金", u"土", u"日"]


def get_calendar_events(calendar):

    empty_flag = True

    for component in calendar.walk():
        if component.name == "VEVENT":
            empty_flag = False
            yield component

    if empty_flag is True:
        raise SystemError("There are not events in this ics file.")


def compute_calendar_data(rawdata, args):

    calendar = Calendar.from_ical(rawdata)
    csv_data = ""

    for event in get_calendar_events(calendar):

        event_start_dt = event.get("dtstart").dt
        event_end_dt = event.get("dtend").dt

        if args.year:

            if event_start_dt.year == args.year:
                csv_data += build_csv_data(event_start_dt, event_end_dt)

        elif args.month:

            if event_start_dt.month == args.month:
                csv_data += build_csv_data(event_start_dt, event_end_dt)

        elif args.range_from and args.range_to:

            if args.range_from < event_start_dt < args.range_to:
                csv_data += build_csv_data(event_start_dt, event_end_dt)

        else:
            csv_data += build_csv_data(event_start_dt, event_end_dt)

    return csv_data


def build_csv_data(start_dt, end_dt):

    template = u"{date},{weekday},{start},~,{end},"
    template += u"{hours}時間{minutes}分,{total},{basetime}\n"

    data = {}
    data["date"] = start_dt.strftime(DATE_FORMAT)
    data["weekday"] = WEEKDAYS[start_dt.weekday()]
    data["start"] = start_dt.strftime(TIME_FORMAT)
    data["end"] = end_dt.strftime(TIME_FORMAT)

    time_delta = relativedelta(end_dt, start_dt)
    data["hours"] = time_delta.hours
    data["minutes"] = time_delta.minutes

    decimal_minutes = float(data["minutes"]) / 60
    decimal_totaltime = round((float(data["hours"]) + decimal_minutes), 2)
    data["total"] = decimal_totaltime

    data["basetime"] = round((decimal_totaltime - 0.5), 2)

    return template.format(**data)


icalwage/utils.py
#!/usr/bin/env python
# coding: utf-8

import os

from pytz import timezone
from datetime import datetime, timedelta


def _get_datetime(query):

    BASIC_ERROR = "Please set full date. ok: '2017.9.1' no: '9.1'"

    fulldate_data = query.split(".")

    if len(fulldate_data) != 3:
        raise TypeError(BASIC_ERROR)

    year, month, day = fulldate_data

    try:
        return datetime(
            int(year), int(month), int(day),
            tzinfo=timezone('Asia/Tokyo')
        )

    except:
        raise TypeError("Incorrect date. " + BASIC_ERROR)


def adjust_args(args):

    if args.year and args.month:
        raise SystemError(
            "You can't set '--year' and '--month' argument by together.")
    if (args.range_from and not args.range_to) or \
       (not args.range_from and args.range_to):
        raise SystemError(
            "Please set '--from' and '--to' argument by together.")

    if args.range_from:

        args.range_from = _get_datetime(args.range_from)

    if args.range_to:

        args.range_to = _get_datetime(args.range_to)
        args.range_to += timedelta(days=1)

    if args.year:

        try:
            args.year = int(args.year)

        except:
            raise TypeError("Please set christian era. ok: '1989' no: '89'")

    if args.month:

        try:
            args.month = int(args.month)

        except:
            raise TypeError("Please set integer. ok: '6' no: 'June'")

    return args


def check_destfile(destfile, overwrite=False):

    if os.access(destfile, os.F_OK):

        if overwrite:
            return True

        else:
            raise IOError(
                "The dest file is already exist. If you allow overwrite, "
                "use '--overwrite' argument.")

    else:
        return True


def check_sourcefile(sourcefile):

    if not sourcefile.endswith(".ics"):
        raise IOError("The source file is not '.ics' file!")

    else:
        return True


icalwage/__main__.py
#!/usr/bin/env python
# coding: utf-8

import sys
import codecs

from .cli import parser
from .core import compute_calendar_data
from .utils import (
    adjust_args,
    check_destfile,
    check_sourcefile
)


def program(args):

    args = adjust_args(args)
    check_sourcefile(args.source)
    check_destfile(args.dest, overwrite=args.overwrite)

    with open(args.source, "r") as sfp:

        context = sfp.read()

    with codecs.open(args.dest, "w", "shift_jis") as dfp:

        csv_data = compute_calendar_data(context, args)
        dfp.write(csv_data)

    print "Created {}".format(args.dest)
    return 0


def main(argv=sys.argv):

    if len(argv) == 1:
        parser.parse_args(["-h"])
        sys.exit(1)

    args = parser.parse_args()

    try:
        status = program(args)
        sys.exit(status)

    except Exception as e:
        error_type = type(e).__name__
        sys.stderr.write("{0}: {1}\n".format(error_type, e.message))
        sys.exit(1)


if __name__ == "__main__":

    main()


icalwage/cli.py
#!/usr/bin/env python
# coding: utf-8

from argparse import ArgumentParser

parser = ArgumentParser(
    prog="icalwage.py",
    description="Load calender data from an .ics file, and Output .csv file.")

# POSITIONALS
parser.add_argument(
    "source",
    metavar="ICS FILE",
    action="store",
    help="The '.ics' file, this file needed to load work informations.")

parser.add_argument(
    "dest",
    metavar="CSV FILE",
    action="store",
    help="A file name contains csv data.")

# OPTIONALS
parser.add_argument(
    "-o", "--overwrite",
    dest="overwrite",
    action="store_true",
    help="If dest csv file is already exist, overwrite that file.")

parser.add_argument(
    "--month",
    dest="month",
    action="store",
    help="Returns the month data only, use this when ics file is heavy. "
         "Please set integer not string. ok: '6' no: 'June'")

parser.add_argument(
    "--year",
    dest="year",
    action="store",
    help="Returns the year data only. Please set christian era. "
         "ok: '1989' no: '89'")

parser.add_argument(
    "--from",
    metavar="FULL DATE",
    dest="range_from",
    action="store",
    help="If you want to limit the range of a date, use this argument "
         "with '--to' arg. Please set full date. ok: '2017.9.1' no: '9.1'")

parser.add_argument(
    "--to",
    metavar="FULL DATE",
    dest="range_to",
    action="store",
    help="If you set '--from' argument, please set this argument.")
